hips/hip-9999.md

```diff
@@ -0,0 +1,69 @@
+---
+hip: 9999
+title: "Add Repository URL to a charts release metadata"
+authors: [ "Luke Reed <luke@lreed.net>" ]
+created: "2021-11-24"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+Currently there is no way of knowing the exact repository that a chart originated from for a given release. There are ways to guess by looking at the sources list in the chart metadata, but since this is provided via the `Chart.yaml` file, any forks of that repository will look identical. The only way to concretely know what the repository URL is that was used at install/upgrade time is to populate a repository URL metadata field in the release during the install or upgrade. It should work for both traditional repositories behind an http web server, as well as an OCI compliant repository.
+
+## Motivation
+
+The primary motivation is to have a helm-native way to track where a chart came from. Many of the motivations for this feature can be found in this issue: https://github.com/helm/helm/issues/4256. To note a couple:
+- An admin would like to know where a chart came from only by querying the cluster itself. Currently the only way to know where a chart came from is to keep track of it externally.
+- Third-party tooling to that provides recommendations on helm upgrades can query the cluster to get the repository URL to aid in this recommendation. See [nova](https://github.com/FairwindsOps/nova) as an example.
+- This metadata field could also populate output for the `helm ls` command in the future. That may be outside the scope of this particular feature.
+
+## Rationale
+
+In order to query for this in the cluster, the repository URL will need to be somewhere in the release object. The chart metadata portion of the release object already contains similar data and makes the most sense for this purpose.
+
+## Specification
+
+Nothing would change from the user's perspective. With some modifications in the current code it's possible to add this metadata into the release object transparently.
+
+The current idea for this implementation is to add a field in both the `ChartDownloader` and `Metadata` (in the `pkg/chart/metadata.go` file) structs to hold the repository URL. The Downloader is the logical place to derive the proper URL and can then that storage location can populate the Metadata field at install/upgrade time.
+
+The chart metadata object would have a field that looked like so:
+
+```
+"repoURL": "https://charts.fairwinds.com/stable"
+```
+
+And an OCI compliant repository would look like this (example is using ECR):
+
+```
+"repoURL": "oci://${ACCOUNT_NUMBER}.dkr.ecr.us-east-1.amazonaws.com/goldilocks"
+```
+
+## Backwards compatibility
+
+There should be no impact to backwards compatibility.
+
+## Security implications
+
+There should be no security implications for this feature.
+
+## How to teach this
+
+N/A, the primary audience for this would be cluster administrators or third-party helm related tool developers.
```

Comment:
“Teaching” can also mean “what documentation should I write to describe this feature?”

I think some thought needs to be put here to describe the feature, how and where the URL is generated, and how to interact with the URL (or if it’s purely informational).

Reply:
Sure, I can add some information into the HIP about this.